design.md

```diff
@@ -0,0 +1,109 @@
+* The worker library provides methods to start, stop and get the status of a running job.
+A method is also provided to stream the output of a job. The api can also be queried for all previous output of a running job. The server also provides a stream for clients to listen to job output.
+* The grpc api is a grpc service that sits on top of the worker library. It provides methods to start, stop and get the status of a running job remotely. 
+* The client/cli is an interface for users to start, stop and get the status of a running job on a remote server.
+
+## Worker library API
+The worker library consists of 2 components. The supervisor and the process.
+### Supervisor API
+The supervisor is responsible for managing all jobs in the system. It provides apis to start, stop, get the output and get the status of a job. 
+* `Newsupervisor()` - Returns a new instance of a supervisor
+* `Start(cmd string, args []string, opts...Options)` - Starts a process and return a unique job id. The job id is generated by the supervisor.
+* `Stop(id string)` - Stops a running process by sending the process group a `SIGTERM` signal.
+* `Run()` - Run the supervisor and listen for incoming job requests.
+* `WatchOutput(id string, c chan ProcessOutput)` - Add an output channel to subscribe to process output. All previous output, if any, from the process is replayed before sending new messages to the channel.
+* `UnWatchOutput(id string, c chan ProcessOutput)` - Unsubscribe to the process output.
+* `GetOutput(id string)` - Get all lines of output from the job's line buffer. Useful when getting the output of a finished job.
+* `Status(id string)` - Returns the status of a process
+
+
+## Process API
+`Process` is a wrapper around `os/exec.Cmd`
+* `NewProcess(name string, args []string, opts ...Options)` - Returns a new instance of a process given a command name and it's arguments. An optional list of options can also be specified.
+* `Start()` - Starts the process and returns a channel the caller can listen on to get the final status.
+* `Stop()` - Stops the process by sending a `SIGTERM` signal to it's process group.
+* `Signal(sig syscall.Signal)` - Send an os signal to the process group.
+* `Status()` - Get the current status of a process.
+* `IsInitialState()` - true if the process is in the initial state (the process isn't started yet).
+* `IsRunningState()` - true if the process is starting or running.
+* `IsFinalState()` - true if the process is in a final state
+
+### Output streaming
+Each process has a `StreamOutput` for managing process output. `StreamOutput` is an `io.Writer` that has a channel where new lines of output from the process are sent to. The caller should use this channel to listen for output from the process. The supervisor also stores each line of output comming from the `StreamOutput` for each job. These can be retrieved at any time by calling `GetOutput(...)`
```

Comment:
How will this work when multiple clients call `WatchOutput` on the same job?

Reply:
`WatchOutput` adds the passed in channel to a set of channels for that job. When the supervisor receives a line of output from a job, it loops through the set of channels for that job and sends the line.